File: android/app/src/main/java/uk/co/darkerwaters/scorepal/data/ScoreState.java
package uk.co.darkerwaters.scorepal.data;

import java.util.ArrayList;
import java.util.List;

public class ScoreState {

    private int currentState;
    private MatchSetup.Team teamChanged;
    private int levelChanged;

    ScoreState() {
        reset();
    }

    public enum ScoreChange {
        NONE(0),
        INCREMENT(1),
        SERVER(2),
        ENDS(4),
        DECIDING_POINT(8),
        TIE_BREAK(16),
        DECREMENT(32),
        BREAK_POINT_CONVERTED(64),
        BREAK_POINT(128),
        INCREMENT_REDO(256);

        final int val;
        ScoreChange(int val) {
            this.val = val;
        }
    }

    public void reset() {
        this.currentState = 0;
        this.teamChanged = null;
        this.levelChanged = -1;
    }

    public boolean isEmpty() { return this.currentState == ScoreChange.NONE.val; }

    public void addChange(ScoreChange change) {
        currentState |= change.val;
    }

    public void addChange(ScoreChange change, MatchSetup.Team team, int level) {
        addChange(change);
        teamChanged = team;
        levelChanged = Math.max(level, levelChanged);
    }

    public static boolean Changed(int state, ScoreChange change) {
        return 0 != (state & change.val);
    }

    public boolean isChanged(ScoreChange change) {
        return 0 != (currentState & change.val);
    }

    public List<ScoreChange> getChanges() {
        // find all the activiated changes and return as a list
        List<ScoreChange> list = new ArrayList<>();
        for (ScoreChange change: ScoreChange.values()) {
            if (isChanged(change)) {
                list.add(change);
            }
        }
        return list;
    }

    public int getState() {
        return currentState;
    }

    public MatchSetup.Team getTeamChanged() { return teamChanged; }
    public int getLevelChanged() { return levelChanged; }

    public void setState(int state, int levelChanged, MatchSetup.Team teamChanged) {
        this.currentState = state;
        this.levelChanged = levelChanged;
        this.teamChanged = teamChanged;
    }
}

File: multiphone/lib/match/score_state.dart
import 'dart:math';

import 'package:multiphone/providers/active_setup.dart';

class ScoreChange {
  static final none = ScoreChange(0);
  static final increment = ScoreChange(1);
  static final server = ScoreChange(2);
  static final ends = ScoreChange(4);
  static final decidingPoint = ScoreChange(8);
  static final tieBreak = ScoreChange(16);
  static final decrement = ScoreChange(32);
  static final breakPointConverted = ScoreChange(64);
  static final breakPoint = ScoreChange(128);
  static final incrementRedo = ScoreChange(256);

  final int val;
  static final List<ScoreChange> values = [
    none,
    increment,
    server,
    ends,
    decidingPoint,
    tieBreak,
    decrement,
    breakPointConverted,
    breakPoint,
    incrementRedo,
  ];

  ScoreChange(int val) : val = val;
}

class ScoreState {
  int _currentState;
  TeamIndex _teamChanged;
  int _levelChanged;

  ScoreState() {
    reset();
  }

  void reset() {
    _currentState = 0;
    _teamChanged = null;
    _levelChanged = -1;
  }

  get isEmpty {
    return _currentState == ScoreChange.none.val;
  }

  void addStateChange(ScoreChange change) {
    _currentState |= change.val;
  }

  void addChange(ScoreChange change, TeamIndex team, int level) {
    addStateChange(change);
    _teamChanged = team;
    _levelChanged = max(level, _levelChanged);
  }

  static bool changed(int state, ScoreChange change) {
    return 0 != (state & change.val);
  }

  bool isChanged(ScoreChange change) {
    return 0 != (_currentState & change.val);
  }

  List<ScoreChange> getChanges() {
    // find all the activated changes and return as a list
    List<ScoreChange> list = [];
    for (int i = 0; i < ScoreChange.values.length; ++i) {
      ScoreChange change = ScoreChange.values[i];
      if (isChanged(change)) {
        list.add(change);
      }
    }
    return list;
  }

  int getState() {
    return _currentState;
  }

  TeamIndex getTeamChanged() {
    return _teamChanged;
  }

  int getLevelChanged() {
    return _levelChanged;
  }

  void setState(int state, int levelChanged, TeamIndex teamChanged) {
    _currentState = state;
    _levelChanged = levelChanged;
    _teamChanged = teamChanged;
  }
}
